Firestore/core/test/unit/bundle/bundle_serializer_test.cc

```diff
@@ -72,51 +245,825 @@ TEST(BundleSerializerTest, DecodesBundleMetadata) {
   EXPECT_EQ(proto_metadata.total_documents(), actual.total_documents());
 }
 
-TEST(BundleSerializerTest, DecodesInvalidBundleMetadataReportsError) {
+TEST_F(BundleSerializerTest, DecodesInvalidBundleMetadataReportsError) {
   auto proto_metadata = TestBundleMetadata();
 
   std::string json_string;
   MessageToJsonString(proto_metadata, &json_string);
   auto invalid = "123" + json_string;
 
-  BundleSerializer serializer;
-  ReadContext context;
-  serializer.DecodeBundleMetadata(context, invalid);
+  JsonReader reader;
+  bundle_serializer.DecodeBundleMetadata(reader, invalid);
 
-  EXPECT_FALSE(context.ok());
+  EXPECT_FALSE(reader.ok());
 
   // Replace total_bytes to a string unparseable to integer.
   std::string json_copy =
       ReplacedCopy(json_string, "123456789987654321", "xxxyyyzzz");
-  context = ReadContext{};
-  EXPECT_TRUE(context.ok());
-  serializer.DecodeBundleMetadata(context, json_copy);
+  reader = JsonReader{};
```

Comment:
This reassignment between cases is something of a smell. There's a bunch of essentially unrelated cases together in this test that are sharing state. You can avoid unintentionally bleeding through from one case to the other by creating separate blocks for each one. e.g.

```c++
{
  JSONReader reader;
  bundle_serializer.DecodeBundleMetadata(reader, invalid);
  EXPECT_NOT_OK(reader.status());
}

{
  std::string json_copy =
      ReplacedCopy(json_string, "123456789987654321", "xxxyyyzzz");
  JsonReader reader;
  bundle_serializer.DecodeBundleMetadata(reader, json_copy);
  EXPECT_NOT_OK(reader.status());
}
```

This keeps any local changes and other objects scoped to each sub-case.

Reply:
Done.